### PR TITLE
Added esp service config rollouts info to /endpoints_status

### PR DESCRIPTION
**Pull request by mangchiandjjoe**
This change will add espRollouts to the output of  /endpoints_status

 "espRollouts": [
  {
   "serviceName": "quota.jaebonginternal.hamiltonparkdr.com",
   "rolloutId": "2017-07-18r19",
   "percentages": {
    "2017-07-18r19": 100
   }
  }
 ]

### CODE REVIEW

File: src/api_manager/BUILD [hunk 1]
```diff
@@ -302,20 +302,19 @@ cc_test(
...
-#         "config_manager.h",
-#         "config_manager_test.cc",
-#     ],
-#     deps = [
-#         ":api_manager",
-#         ":mock_api_manager_environment",
-#         "//external:googletest_main",
-#     ],
-# )
+cc_test(
```
Comment (qiwzhang) [inline src/api_manager/BUILD]: this should not be part of this PR

Comment (mangchiandjjoe) [inline src/api_manager/BUILD]: That is not part of the PR. These are changed files included .

$ git status
On branch master
Your branch is ahead of 'origin/master' by 2 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   include/api_manager/api_manager.h
	modified:   src/api_manager/api_manager_impl.cc
	modified:   src/api_manager/api_manager_impl.h
	modified:   src/api_manager/weighted_selector.h
	modified:   src/nginx/proto/status.proto
	modified:   src/nginx/status.cc

It seems that meringue back from master included old commits, but those are not part of this.

File: src/nginx/proto/status.proto [hunk 2]
```diff
@@ -106,11 +106,22 @@ message ProcessStatus {
...
 // Top-level endpoints status message
 message Status {
   // Overall server status
   ServerStatus server = 1;
 
   // Status for each process
   repeated ProcessStatus processes = 2;
+
+  // ESP rollouts
+  repeated ServiceConfigRollouts esp_rollouts = 3;
```
Comment (lizan) [inline src/nginx/proto/status.proto]: why this is out of ProcessStatus?
Isn't each worker process have different cycle to reload rollouts?

Comment (qiwzhang) [inline src/nginx/proto/status.proto]: That is true.  It is possible that different worker (process) may use different rollout id since they refresh independently. 
But put them in ProcessStatus is tricky,  it is stored in share memory and updated by a timer.
If we assume everything works,   getting rollout id just from one worker is good enough.

Comment (lizan) [inline src/nginx/proto/status.proto]: I know the trickiness to put it into ProcessStatus, but at we have been using /endpoints_status for debug purposely, IMO we shouldn't assume everything works.

Comment (mangchiandjjoe) [inline src/nginx/proto/status.proto]: esp module doesn't  have the list of the api_manager references in the shared memory.
To move this information into ProcessStatus, we need to keep those information into the share memory.
Let me investigate more.

Comment (mangchiandjjoe) [inline src/nginx/proto/status.proto]: We can add the fixed length char array buffer to the share memory and store the serialized rollouts information. I think this works in most case.

<rollout_id>|<config_id count>|<config_id 1>|<percentage 1>|<config_id 2>|<percentage 2>|...|<config_id n>|<percentage n>\0

This information is updated initially and when a new rollout deployed.

Comment (qiwzhang) [inline src/nginx/proto/status.proto]: Yes,  all string ids are fixed length,  we can define a max_config as 10.  so the size is fixed. 
In the timer,  only  write to shared memory when data is changed.  so to reduce chances of race condition (reading and writing in different process)

Comment (lizan) [inline src/nginx/proto/status.proto]: Sounds good to me. Also you might be able to just update the shared memory when you fetched new rollouts.

**Comment by mangchiandjjoe** [comment]
{
"server": {
"serverVersion": "ESP/1.6.0",
"nginxServerVersion": "nginx/1.13.3",
"builtBy": "gcc 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04.3) ",
"configure": " --builddir=bazel-out/local-fastbuild/genfiles/external/nginx/bazel-objs --with-cc=/usr/bin/gcc --with-cc-opt= --prefix=/etc/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --user=nginx --group=nginx --http-log-path=/var/log/nginx/access.log --http-client-body-temp-path=/var/cache/nginx/client_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --without-http --without-http-cache --without-http_auth_basic_module --without-http_upstream_zone_module",
"connections": {
"active": "2",
"accepted": "2",
"handled": "2",
"reading": "0",
"writing": "1",
"waiting": "1"
},
"requests": "1"
},
"processes": [
{
"processId": "167578",
"memoryUsage": "5885952",
"peakMemoryUsage": "5885952",
"virtualSize": "33533952",
"userCpuTimeUs": "0",
"systemCpuTimeUs": "0",
"espStatus": [
{
"serviceName": "quota.jaebonginternal.hamiltonparkdr.com",
"serviceControlStatistics": {
"totalCalledChecks": "0",
"sendChecksByFlush": "0",
"sendChecksInFlight": "0",
"totalCalledReports": "0",
"sendReportsByFlush": "0",
"sendReportsInFlight": "0",
"sendReportOperations": "0",
"maxReportSize": "0"
},
"serviceConfigRollouts": {
"rolloutId": "2017-05-16r1",
"percentages": {
"2017-05-16r1": "100"
}
}
}
],
"startedAt": "2017-07-20T17:37:02Z"
}
]
}

File: src/nginx/status.cc [hunk 3]
```diff
@@ -317,6 +317,15 @@ ngx_int_t ngx_esp_init_process_stats(ngx_cycle_t *cycle) {
   for (ngx_uint_t i = 0, napis = mc->endpoints.nelts; i < napis; i++) {
     ngx_esp_loc_conf_t *lc = endpoints[i];
     if (lc->esp) {
+      std::string rollouts;
+      ngx_esp_update_rollout(lc->esp, &rollouts);
+      if (rollouts.length() <= kMaxServiceRolloutsInfoSize) {
+        process_stat->esp_stats[process_stat->num_esp].rollouts_length =
+            rollouts.length();
+        strncpy(process_stat->esp_stats[process_stat->num_esp].rollouts,
```
Comment (qiwzhang) [inline src/nginx/status.cc]: can we compare first, if changed, update share memory.  For share memory, less update is better.

File: src/nginx/status.cc [hunk 4]
```diff
@@ -347,6 +356,14 @@ ngx_int_t ngx_esp_init_process_stats(ngx_cycle_t *cycle) {
     for (ngx_uint_t i = 0, napis = mc->endpoints.nelts; i < napis; i++) {
       ngx_esp_loc_conf_t *lc = endpoints[i];
       if (lc->esp) {
+        std::string rollouts;
+        ngx_esp_update_rollout(lc->esp, &rollouts);
+        if (rollouts.length() <= kMaxServiceRolloutsInfoSize) {
+          process_stat->esp_stats[esp_idx].rollouts_length = rollouts.length();
```
Comment (qiwzhang) [inline src/nginx/status.cc]: please try to re-use code,  it seems it is the same as line 321

File: src/nginx/status.cc [hunk 5]
```diff
@@ -125,6 +128,9 @@ void fill_process_stats(const ngx_esp_process_stats_t &stat,
     fill_service_control_statistics(
         stat.esp_stats[j].statistics.service_control_statistics,
         esp_status_proto->mutable_service_control_statistics());
+    esp_status_proto->mutable_service_config_rollouts()->ParseFromString(
```
Comment (lizan) [inline src/nginx/status.cc]: Use `ParseFromArray` instead, then you don't need to construct a `std::string`

File: src/nginx/status.h [hunk 6]
```diff
@@ -41,6 +41,7 @@ namespace nginx {
 // The maximum number of esp objects.
 const int kMaxEspNum = 10;
 const int kMaxServiceNameSize = 256;
+const int kMaxServiceRolloutsInfoSize = 256;
```
Comment (lizan) [inline src/nginx/status.h]: This looks like too small to me, do you have idea roughly how many config_ids will fit into this size?

File: src/nginx/status.cc [hunk 7]
```diff
@@ -270,6 +276,34 @@ Status stats_json_per_process(const ngx_esp_process_stats_t &process_stats,
...
+
+  std::string rollouts;
+  service_config_rollouts_proto.SerializeToString(&rollouts);
+
+  if (rollouts.length() <= kMaxServiceRolloutsInfoSize &&
+      rollouts.length() > 0 &&
+      strncmp(process_stat->esp_stats[index].rollouts, rollouts.c_str(),
+              rollouts.length()) != 0) {
+    process_stat->esp_stats[index].rollouts_length = rollouts.length();
+    strncpy(process_stat->esp_stats[index].rollouts, rollouts.c_str(),
```
Comment (lizan) [inline src/nginx/status.cc]: use `memcpy`, binary serialized protobuf might have NULL in the middle. Also consider check `ByteSize` of protobuf and directly use `SerializeToArray` instead of copying.

File: src/nginx/status.h (hunk 6 above)
Comment (mangchiandjjoe) [inline src/nginx/status.h]: As Wayne mentioned, we can define the maximum number of config_id in a rollout to 10. 

According to my test, if the config_id is generated by the server, 256 seems to be good enough.
But config_id can be assigned by user and maximum length of it is not defined in the service.proto, we might not be able to set long enough.

I increased this to 4096.

File: src/nginx/status.cc (hunk 7 above)
Comment (mangchiandjjoe) [inline src/nginx/status.cc]: Used memcmp, memcpy, ByteSize and SerializeToArray

File: src/nginx/status.cc (hunk 5 above)
Comment (mangchiandjjoe) [inline src/nginx/status.cc]: Done

File: src/nginx/status.cc (hunk 7 above)
Comment (mangchiandjjoe) [inline src/nginx/status.cc]: It seems that memcmp doesn't work. Output of SerializeToArray doesn't follow the order added.

Comment (mangchiandjjoe) [inline src/nginx/status.cc]: Because of item orders in a map, wrote map compare to check update

Comment (lizan) [inline src/nginx/status.cc]: hmmm, that might be overkill, try MessageDifferencer in protobuf::util.

Comment (mangchiandjjoe) [inline src/nginx/status.cc]: It's good to know. Used MessageDifferencer to compare protobuf instances

**Comment by mangchiandjjoe** [comment]
I added related tests to existing config_rollouts_managed.t

File: src/nginx/status.cc [hunk 8]
```diff
@@ -26,9 +26,9 @@
 //
 #include "src/nginx/status.h"
 
+#include <google/protobuf/util/message_differencer.h>
```
Comment (lizan) [inline src/nginx/status.cc]: nit: use `"` for non system includes

File: src/nginx/t/config_rollouts_managed.t [hunk 9]
```diff
@@ -154,6 +159,18 @@ is( $response_body, <<'EOF', 'Shelves returned in the response body.' );
 EOF
 
 
+# verify the first /endpoints_status
+$response = ApiManager::http_get($NginxPort, '/endpoints_status' );
+( $response_headers, $response_body ) = split /\r\n\r\n/, $response, 2;
+like( $response_headers, qr/HTTP\/1\.1 200 OK/, 'Returned HTTP 200.' );
+my $endpoints_status = decode_json( $response_body );
```
Comment (lizan) [inline src/nginx/t/config_rollouts_managed.t]: you might be able to use `ApiManager::compare_json` to do these?

File: src/nginx/status.cc (hunk 8 above)
Comment (mangchiandjjoe) [inline src/nginx/status.cc]: Done

File: src/nginx/t/config_rollouts_managed.t (hunk 9 above)
Comment (mangchiandjjoe) [inline src/nginx/t/config_rollouts_managed.t]: Output of the /endpoints_status contains  system and process statuses data and statistics.
Most of data can be different from each request.
We only need to compare serviceConfigRollouts.

File: src/nginx/status.cc [hunk 10]
```diff
@@ -270,6 +276,38 @@ Status stats_json_per_process(const ngx_esp_process_stats_t &process_stats,
...
+
+  int length = new_rollouts.ByteSize();
+  if (0 < length && length <= kMaxServiceRolloutsInfoSize) {
+    char rollouts[kMaxServiceRolloutsInfoSize];
+    new_rollouts.SerializeToArray(rollouts, length);
+
+    ServiceConfigRolloutsProto current_rollouts;
+    if (current_rollouts.ParseFromArray(
+            process_stat->esp_stats[index].rollouts,
+            process_stat->esp_stats[index].rollouts_length)) {
```
Comment (qiwzhang) [inline src/nginx/status.cc]: does ParseFromArray() return true of size is 0?   

If fail, memory is corrupted, we may want to override it

File: src/nginx/status.cc [hunk 11]
```diff
@@ -270,6 +276,38 @@ Status stats_json_per_process(const ngx_esp_process_stats_t &process_stats,
...
+      if (!google::protobuf::util::MessageDifferencer::Equals(current_rollouts,
+                                                              new_rollouts)) {
+        memcpy(process_stat->esp_stats[index].rollouts, rollouts, length);
```
Comment (qiwzhang) [inline src/nginx/status.cc]: move line 294 here

File: src/nginx/status.cc (hunk 10 above)
Comment (lizan) [inline src/nginx/status.cc]: I think so.

Comment (mangchiandjjoe) [inline src/nginx/status.cc]: ParseFromArray returns true if the length is 0.

Added error recovery code.

File: src/nginx/status.cc (hunk 11 above)
Comment (mangchiandjjoe) [inline src/nginx/status.cc]: Moved
